satellite/satellitedb/overlaycache.go

```diff
@@ -87,41 +103,91 @@ func (cache *overlaycache) SelectNewStorageNodes(ctx context.Context, count int,
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range nodes {
+			totalNodes = append(totalNodes, n)
+			criteria.ExcludedNodes = append(criteria.ExcludedNodes, n.Id)
+			criteria.ExcludedIPs = append(criteria.ExcludedIPs, n.LastIp)
+		}
+		if len(totalNodes) == count {
+			break
+		}
+	}
+	return totalNodes, nil
+}
+
+func (cache *overlaycache) queryFilteredNodes(ctx context.Context, excludedNodes []storj.NodeID, excludedIPs []string, count int, safeQuery string, distinctIP bool, args ...interface{}) (_ []*pb.Node, err error) {
+	switch t := cache.db.DB.Driver().(type) {
+	case *sqlite3.SQLiteDriver:
+		return cache.sqliteQueryNodes(ctx, excludedNodes, excludedIPs, count, safeQuery, distinctIP, args...)
+	case *pq.Driver:
+		return cache.postgresQueryNodes(ctx, excludedNodes, excludedIPs, count, safeQuery, distinctIP, args...)
+	default:
+		return []*pb.Node{}, Error.New("Unsupported database %t", t)
+	}
 }
 
-func (cache *overlaycache) queryFilteredNodes(ctx context.Context, excluded []storj.NodeID, count int, safeQuery string, args ...interface{}) (_ []*pb.Node, err error) {
+func (cache *overlaycache) sqliteQueryNodes(ctx context.Context, excludedNodes []storj.NodeID, excludedIPs []string, count int, safeQuery string, distinctIP bool, args ...interface{}) (_ []*pb.Node, err error) {
 	if count == 0 {
 		return nil, nil
 	}
 
 	safeExcludeNodes := ""
-	if len(excluded) > 0 {
-		safeExcludeNodes = ` AND id NOT IN (?` + strings.Repeat(", ?", len(excluded)-1) + `)`
+	if len(excludedNodes) > 0 {
+		safeExcludeNodes = ` AND id NOT IN (?` + strings.Repeat(", ?", len(excludedNodes)-1) + `)`
+		for _, id := range excludedNodes {
+			args = append(args, id.Bytes())
+		}
 	}
-	for _, id := range excluded {
-		args = append(args, id.Bytes())
+
+	safeExcludeIPs := ""
+	if len(excludedIPs) > 0 {
+		safeExcludeIPs = ` AND last_ip NOT IN (?` + strings.Repeat(", ?", len(excludedIPs)-1) + `)`
+		for _, ip := range excludedIPs {
+			args = append(args, ip)
+		}
 	}
+
 	args = append(args, count)
 
-	rows, err := cache.db.Query(cache.db.Rebind(`SELECT id,
-		type, address, free_bandwidth, free_disk, audit_success_ratio,
+	var rows *sql.Rows
+	if distinctIP {
+		rows, err = cache.db.Query(cache.db.Rebind(`SELECT id,
+		type, address, last_ip, free_bandwidth, free_disk, audit_success_ratio,
+		uptime_ratio, total_audit_count, audit_success_count, total_uptime_count,
+		uptime_success_count
+		FROM (SELECT id, type, address, last_ip, free_bandwidth, free_disk, audit_success_ratio,
+			uptime_ratio, total_audit_count, audit_success_count, total_uptime_count, uptime_success_count,
+			Row_number() OVER(PARTITION BY last_ip ORDER BY RANDOM()) rn
+			FROM nodes
+			`+safeQuery+safeExcludeNodes+safeExcludeIPs+`) n
+		WHERE rn = 1
+		ORDER BY RANDOM()
+		LIMIT ?`), args...)
+	} else {
+		rows, err = cache.db.Query(cache.db.Rebind(`SELECT id,
```

Comment:
I feel like we should move the if distinctIP Check outside of the function and have the "normal" query being a seperate function.
Ala 
* QueryNodes
* PostgresQueryNodesDistinct
* SQLiteQueryNodesDistinct

Based on the bool we then can use the same query for  SQLite and Postgres and only switch based on the driver if we want distinct nodes.